test/performance/scale_from_zero_test.go

```diff
@@ -0,0 +1,198 @@
+// +build performance
+
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package performance
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"testing"
+	"time"
+
+	pkgTest "github.com/knative/pkg/test"
+	"github.com/knative/pkg/test/logging"
+
+	"github.com/knative/serving/test"
+	"k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	configName = "perftest-config"
+	routeName  = "perftest-route"
+)
+
+const (
+	helloWorldExpectedOutput = "Hello World! How about some tasty noodles?"
+)
+
+type testContext struct {
+	logger         *logging.BaseLogger
+	deploymentName string
+	names          test.ResourceNames
+	domain         string
+}
+
+func setupClients(t *testing.T) *test.Clients {
+	clients, err := test.NewClients(
+		pkgTest.Flags.Kubeconfig,
+		pkgTest.Flags.Cluster,
+		test.ServingNamespace)
+	if err != nil {
+		t.Fatalf("Couldn't initialize clients: %v", err)
+	}
+	return clients
+}
+
+func tearDown(clients *test.Clients, names test.ResourceNames, logger *logging.BaseLogger) {
+	if clients != nil && clients.ServingClient != nil {
+		clients.ServingClient.Delete([]string{names.Route}, []string{names.Config}, []string{names.Service})
+	}
+}
+
+func createRouteAndConfig(clients *test.Clients, logger *logging.BaseLogger, imagePath string, options *test.Options) (test.ResourceNames, error) {
+	var names test.ResourceNames
+	names.Config = test.AppendRandomString(configName, logger)
+	names.Route = test.AppendRandomString(routeName, logger)
+
+	if err := test.CreateConfiguration(logger, clients, names, imagePath, options); err != nil {
+		return test.ResourceNames{}, err
+	}
+	err := test.CreateRoute(logger, clients, names)
+	return names, err
+}
+
+func isDeploymentScaledToZero() func(d *v1beta1.Deployment) (bool, error) {
```

Comment:
(a side comment) - actually it would be great to somehow annotate the deployment so its by default at 0 scale so we don't have to wait at all when we are testing

Reply:
yes, there have been discussions around having such a mechanism in the AutoscaleUpDown test as well. Do you mind opening an issue for a focused discussion on such an annotation?

Reply:
Linking the issue here https://github.com/knative/serving/issues/2328